test/Sema/enum_equatable_hashable.swift

```diff
@@ -89,6 +88,59 @@ enum Complex {
 if Complex.A(1) == .B { } // expected-error{{binary operator '==' cannot be applied to operands of type 'Complex' and '_'}}
 // expected-note @-1 {{overloads for '==' exist with these partially matching parameter lists: }}
 
+// Enums with equatable payloads are equatable if they explicitly conform.
+enum EnumWithEquatablePayload: Equatable {
+  case A(Int)
+  case B(String, Int)
+  case C
+}
+
+if EnumWithEquatablePayload.A(1) == .B("x", 1) { }
+if EnumWithEquatablePayload.A(1) == .C { }
+if EnumWithEquatablePayload.B("x", 1) == .C { }
+
+// Enums with hashable payloads are hashable if they explicitly conform.
+enum EnumWithHashablePayload: Hashable {
+  case A(Int)
+  case B(String, Int)
+  case C
+}
+
+_ = EnumWithHashablePayload.A(1).hashValue
+_ = EnumWithHashablePayload.B("x", 1).hashValue
+_ = EnumWithHashablePayload.C.hashValue
+
+// ...and they should also inherit equatability from Hashable.
+if EnumWithHashablePayload.A(1) == .B("x", 1) { }
+if EnumWithHashablePayload.A(1) == .C { }
+if EnumWithHashablePayload.B("x", 1) == .C { }
+
+// Enums with non-hashable payloads don't derive conformance.
+struct NotHashable {}
+enum EnumWithNonHashablePayload: Hashable { // expected-error 2 {{does not conform}}
+  case A(NotHashable)
+}
+
+// Enums should be able to derive conformances based on the conformances of
+// their generic arguments.
+enum GenericHashable<T: Hashable>: Hashable {
+  case A(T)
+  case B
+}
+if GenericHashable<String>.A("a") == .B { }
+var genericHashableHash: Int = GenericHashable<String>.A("a").hashValue
+
+// But it should be an error if the generic argument doesn't have the necessary
+// constrants to satisfy the conditions for derivation.
```

Comment:
Not a super buggy but I was just reading through this PR and noticed `constrants` -> `constraints` :)